lib/stix2/relationship_objects/relationship.rb

```diff
@@ -7,6 +7,159 @@ class Relationship < Base
       property :target_ref, required: true, coerce: String
       property :start_time, coerce: Time
       property :stop_time, coerce: Time
+
+      def initialize(args)
+        if !args[:relationship_type] && args[:source_ref] && args[:target_ref]
+          objects = DOMAIN_OBJECTS + CYBEROBSERVABLE_OBJECTS
+          source_type = type_by_id(args[:source_ref])
+          target_type = type_by_id(args[:target_ref])
+          relationships = Array(RELATIONSHIP_TYPES.dig(source_type, target_type))
+          relationships += COMMON_RELATIONSHIPS if objects.include?(source_type) && objects.include?(target_type)
+          args[:relationship_type] = relationships.first unless relationships.empty?
+        end
+
+        super(args)
+      end
+
+      private
+
+      def type_by_id(id)
+        id.split("--").first
+      end
+
+      COMMON_RELATIONSHIPS = ["related-to", "derived-from", "duplicate-of"]
```

Comment:
It is pointless to put a constant in the private section in ruby. Constants are always public. You will need `private_constant`.

Reply:
You should also freeze all those constants.